zipline/utils/calendars/trading_calendar.py

```diff
@@ -0,0 +1,843 @@
+
+        Parameters
+        ----------
+        dt: pd.Timestamp
+            The dt for which to check if this exchange is open.
+
+        Returns
+        -------
+        bool
+            Whether the exchange is open on this dt.
+        """
+        return is_open(self.market_opens_nanos, self.market_closes_nanos,
+                       dt.value)
+
+    def next_open(self, dt):
+        """
+        Given a dt, returns the next open.
+
+        If the given dt happens to be a session open, the next session's open
+        will be returned.
+
+        Parameters
+        ----------
+        dt: pd.Timestamp
+            The dt for which to get the next open.
+
+        Returns
+        -------
+        pd.Timestamp
+            The UTC timestamp of the next open.
+        """
+        idx = next_divider_idx(self.market_opens_nanos, dt.value)
+        return self.schedule.market_open[idx].tz_localize('UTC')
+
+    def next_close(self, dt):
+        """
+        Given a dt, returns the next close.
+
+        Parameters
+        ----------
+        dt: pd.Timestamp
+            The dt for which to get the next close.
+
+        Returns
+        -------
+        pd.Timestamp
+            The UTC timestamp of the next close.
+        """
+        idx = next_divider_idx(self.market_closes_nanos, dt.value)
+        return self.schedule.market_close[idx].tz_localize('UTC')
+
+    def previous_open(self, dt):
+        """
+        Given a dt, returns the previous open.
+
+        Parameters
+        ----------
+        dt: pd.Timestamp
+            The dt for which to get the previous open.
+
+        Returns
+        -------
+        pd.Timestamp
+            The UTC imestamp of the previous open.
+        """
+        idx = previous_divider_idx(self.market_opens_nanos, dt.value)
+        return self.schedule.market_open[idx].tz_localize('UTC')
+
+    def previous_close(self, dt):
+        """
+        Given a dt, returns the previous close.
+
+        Parameters
+        ----------
+        dt: pd.Timestamp
+            The dt for which to get the previous close.
+
+        Returns
+        -------
+        pd.Timestamp
+            The UTC timestamp of the previous close.
+        """
+        idx = previous_divider_idx(self.market_closes_nanos, dt.value)
+        return self.schedule.market_close[idx].tz_localize('UTC')
+
+    def next_minute(self, dt):
+        """
+        Given a dt, return the next exchange minute.  If the given dt is not
+        an exchange minute, returns the next exchange open.
+
+        Parameters
+        ----------
+        dt: pd.Timestamp
+            The dt for which to get the next exchange minute.
+
+        Returns
+        -------
+        pd.Timestamp
+            The next exchange minute.
+        """
+        idx = next_divider_idx(self._trading_minutes_nanos, dt.value)
+        return pd.Timestamp(self._trading_minutes_nanos[idx], tz='UTC')
```

Comment:
Any reason we don't index into `self.all_minutes` instead, so we don't have to construct a new timestamp each time?